torchtune/datasets/multimodal/_the_cauldron.py

```diff
@@ -0,0 +1,165 @@
+    from Hugging Face Datasets.
+
+    Args:
+        model_transform (Transform): model-specific transform that takes in a sample dict and applies custom
+            transforms on the keys. The tokenizer used by the model should be encapsulated in the model transform
+            and should operate on the "messages" field. Any model-specific image transforms should operate on
+            the "images" field. The keys returned by the model should be aligned with the
+            expected inputs into the model.
+        subset (str): name of the subset of the dataset to load. See the `dataset card
+            <https://huggingface.co/datasets/HuggingFaceM4/the_cauldron>`_ for options.
+        source (str): path to dataset repository on Hugging Face. For local datasets,
+            define source as the data file type (e.g. "json", "csv", "text") and pass
+            in the filepath in ``data_files``. See `Hugging Face's
+            <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path>`_
+            ``load_dataset`` for more details. Default is ``HuggingFaceM4/the_cauldron``.
+        column_map (Optional[Dict[str, str]]): a mapping to change the expected "images"
+            and "texts" column names to the actual column names in the dataset. Default is None,
+            keeping the default column names.
+        new_system_prompt (Optional[str]): if specified, prepend a system message. This can
+            serve as instructions to guide the model response. Setting this will OVERRIDE any system
+            messages already present in the dataset. Default is None.
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        packed (bool): Whether or not to pack the dataset to ``max_seq_len`` prior to training. Default is False.
+        split (str): ``split`` argument for ``datasets.load_dataset``. You can use this argument to load a subset
+            of a given split, e.g. ``split="train[:10%]"``. Default is "train".
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to ``load_dataset``. See Hugging
+            Face's `API ref <https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset>`_
+            for more details.
+
+    Returns:
+        Union[SFTDataset, PackedDataset]: dataset configured with source data and transform
+
+    Raises:
+        ValueError: If ``packed`` is True and ``max_seq_len`` is not set on the model_transform.
+
+    Example:
+        >>> cauldron_ds = the_cauldron_dataset(model_transform=model_transform, subset="ai2d")
+        >>> for batch in Dataloader(cauldron_ds, batch_size=8):
+        >>>     print(f"Batch size: {len(batch)}")
+        >>> Batch size: 8
+    """
+
+    message_transform = TheCauldronToMessages(
+        train_on_input=train_on_input,
+        column_map=column_map,
+    )
```

Comment:
Should be passing `new_system_prompt`? Also nit but would call it `to_messages` or something, since `model_transform` also involves `Message` but as an input